src/qibosoq/programs/base.py

```diff
@@ -352,18 +352,32 @@ def group_mux_ro(self) -> list:
 
         Example of list:
         [[pulse1, pulse2], [pulse3]]
+
+        Readout pulses are considered to be correctly organized.
         """
         mux_list: List[List[Element]] = []
-        len_last_readout = 0.0
-        for pulse in (pulse for pulse in self.sequence if pulse.type == "readout"):
-            if pulse.start_delay <= len_last_readout and len(mux_list) > 0:
-                # add the pulse to the last multiplexed readout
-                mux_list[-1].append(pulse)
-            else:
-                # add a new multiplexed readout
-                mux_list.append([pulse])
-            len_last_readout = pulse.duration
-
+        group: List[Element] = []
+
+        for pulse in self.sequence:
+            readout = pulse.type != "readout"
+            if readout or pulse.start_delay != 0:
+                if len(group) > 0:
+                    mux_list.append(group)
+                    group = []
+
+            if not readout:
+                group.append(pulse)
```

Comment:
Reading this again, the variable name sounds a bit misleading. I'd propose to negate it
```suggestion
            readout = pulse.type == "readout"
            if (not readout) or pulse.start_delay != 0:
                if len(group) > 0:
                    mux_list.append(group)
                    group = []

            if readout:
                group.append(pulse)
```